pipeline_dp/budget_accounting.py

```diff
@@ -100,7 +101,205 @@ def compute_budgets(self):
             requested_budget.budget.set_eps_delta(eps, delta)
+
+
+@dataclass
+class MechanismSpec:
+    """Specifies the parameters for a mechanism.
+
+    NoiseKind defines the kind of noise distribution.
+    noise is the minimized noise standard deviation.
+    """
+    noise_kind: NoiseKind
+    _noise_standard_deviation: float = None
+
+    @property
+    def noise_standard_deviation(self):
+        """Noise value for the mechanism.
+
+        Raises:
+            AssertionError: The noise value is not calculated yet.
+        """
+        if self._noise_standard_deviation is None:
+            raise AssertionError(
+                "Noise standard deviation is not calculated yet.")
+        return self._noise_standard_deviation
+
+
+@dataclass
+class MechanismSpecInternal:
+    """Stores sensitivity and weight not exposed in MechanismSpec."""
+    sensitivity: float
+    weight: float
+    mechanism_spec: MechanismSpec
+
+
+class PLDBudgetAccountant:
+    """Manages the privacy budget for privacy loss distributions.
+
+    It manages the privacy budget for the pipeline using the
+    Privacy Loss Distribution (PLD) implementation from Google's
+    dp_accounting library.
+    """
+
+    def __init__(self,
+                 total_epsilon: float,
+                 total_delta: float,
+                 pld_discretization: float = 1e-4):
+        """Constructs a PLDBudgetAccountant.
+
+        Args:
+            total_epsilon: epsilon for the entire pipeline.
+            total_delta: delta for the entire pipeline.
+            pld_discretization: `value_discretization_interval` in PLD library.
+                Smaller interval results in better accuracy, but increases running time.
+
+        Raises:
+            ValueError: Arguments are missing or out of range.
+        """
+
+        _validate_epsilon_delta(total_epsilon, total_delta)
+
+        self._total_epsilon = total_epsilon
+        self._total_delta = total_delta
+        self._mechanisms = []
+        self.minimum_noise_std = None
+        self._pld_discretization = pld_discretization
+
+    def request_budget(self,
+                       noise_kind: NoiseKind,
+                       sensitivity: float = 1,
+                       weight: float = 1) -> MechanismSpec:
+        """Request a budget.
+
+        Constructs a mechanism spec based on the parameters.
+        Adds the mechanism to the pipeline for future calculation.
+
+        Args:
+            noise_kind: The kind of noise distribution for the mechanism.
+            sensitivity: The sensitivity for the mechanism.
+            weight: The weight for the mechanism.
+
+        Returns:
+            A "lazy" mechanism spec object that doesn't contain the noise
+            standard deviation until compute_budgets is called.
+        """
+        if noise_kind == NoiseKind.GAUSSIAN and self._total_delta == 0:
+            raise AssertionError("Gaussian delta value must be greater than 0")
```

Comment:
Nit: "The Gaussian mechanism requires the pipeline delta ..."

Explanation: it's slightly more precise and suggests the API user about "the pipeline delta"